File: atmanatic_research/timestamps.py
"""Shared RFC 3339 timestamp parsing for protocol validators.

Every 0.1 wire timestamp MUST include an explicit UTC offset. Naive
timestamps are rejected rather than silently assumed to be UTC.
"""

from __future__ import annotations

from datetime import datetime


class TimestampError(ValueError):
    """Raised when a value is not a valid RFC 3339 timestamp with an explicit offset."""


def parse_rfc3339(value: object, field: str) -> datetime:
    """Parse `value` as an offset-aware RFC 3339 timestamp or raise `TimestampError`."""
    if not isinstance(value, str) or not value.strip():
        raise TimestampError(f"{field} must be a non-empty timestamp string")
    text = value.strip()
    if text.endswith(("Z", "z")):
        text = f"{text[:-1]}+00:00"
    try:
        parsed = datetime.fromisoformat(text)
    except ValueError as error:
        raise TimestampError(f"{field} must be a valid RFC 3339 timestamp") from error
    if parsed.tzinfo is None:
        raise TimestampError(f"{field} must include an explicit UTC offset")
    return parsed

File: interop/reference-ts/src/timestamps.ts
/** Shared RFC 3339 timestamp parsing, mirroring atmanatic_research/timestamps.py.
 * Every 0.1 wire timestamp MUST include an explicit UTC offset; naive
 * timestamps are rejected rather than silently assumed to be UTC.
 */

const RFC3339_PATTERN =
  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$/;

export class TimestampError extends Error {}

/** Validate `value` as an offset-aware RFC 3339 timestamp or throw `TimestampError`. */
export function parseRfc3339(value: unknown, field: string): Date {
  if (typeof value !== "string" || value.trim() === "") {
    throw new TimestampError(`${field} must be a non-empty timestamp string`);
  }
  const text = value.trim();
  if (!RFC3339_PATTERN.test(text)) {
    throw new TimestampError(`${field} must include an explicit UTC offset`);
  }
  const parsed = new Date(text);
  if (Number.isNaN(parsed.getTime())) {
    throw new TimestampError(`${field} must be a valid RFC 3339 timestamp`);
  }
  return parsed;
}
